### ISSUE TITLE
Output incorrect/corrupted with \r in diff content

### ISSUE THREAD

**Issue by mqudsi** [jeffkaufman/icdiff#109]
Is there a way to get icdiff to show unchanged lines on sides? As it currently stands, I was bashing my head against the wall trying to understand why icdiff was not showing me the changes between two files correctly until I understood that only of the two columns displays unchanged data.

If not by default, then at least in `-W` mode?

**Comment by jeffkaufman** [comment]
It sounds like you're asking for what `-W` / `--whole-file` does?  Can you say more about what you want `icdiff --whole-file ...` to do vs what it's actually doing?  Maybe with a screenshot or some simple example files?  For example, here's one of our tests:

```
$ icdiff tests/input-{6,7}.txt --cols=80 --whole-file
tests/input-6.txt                       tests/input-7.txt
a                                       a
b                                       b
c                                       c
d                                       d
e                                       e
f                                       f
g                                       g
h                                       h
i                                       i
j                                       j
k                                       k
l                                       l
                                        m (this is green)
                                        m (this is green)
m                                       m
n                                       n
o                                       o
p                                       p
q                                       q
r                                       r
s                                       s
t                                       t
u                                       u
z                                       z
w                                       w
x                                       x
y                                       y
z                                       z
```

**Comment by mqudsi** [comment]
You know, I suspected there was something else wrong when I couldn't get `-W` to do anything meaningful.

This is file1:
```
[package]
name = "tcpproxy"
version = "0.1.0"
authors = ["Mahmoud Al-Qudsi <mqudsi@neosmart.net>"]
description = "Cross-platform multi-client TCP proxy"
homepage = "https://github.com/neosmart/tcpproxy"
repository = "https://github.com/neosmart/tcpproxy"
readme = "README.md"
keywords = ["proxy", "tcp", "networking"]
categories = ["command-line-utilities", "network-programming"]
license = "MIT"

[dependencies]
getopts = "0.2"
rand = "0.3"
```

This is file2:
```
[package]
name = "srtresync"
version = "0.1.0"
authors = ["Mahmoud Al-Qudsi <mqudsi@neosmart.net>"]

[dependencies]
```

This is a screenshot of `icdiff -W file1 file2`:
![icdiff1](https://user-images.githubusercontent.com/606923/29242651-48c38258-7f56-11e7-89ab-b5bbe340a181.png)

and this is a screenshot of `icdiff file1 file` (no `-W`):
![icdiff2](https://user-images.githubusercontent.com/606923/29242654-53d2d4a0-7f56-11e7-8747-2d6af1f764bc.png)

**Comment by jeffkaufman** [comment]
Something else is going wrong with your setup, but I'm not sure what.  Here's what that looks like for me, and what it should look like:

<img width="838" alt="icdiff" src="https://user-images.githubusercontent.com/262566/29243974-c2168226-7f7a-11e7-97c1-92a3dae296b9.png">

What happens if you specifically pass in `--cols N` to tell it how many columns to use?

**Comment by mqudsi** [comment]
OK, so at this point seeing as it's more of a bug/problem report and not a feature request, I should probably take a step back and explain that this is on WSL, so basically Ubuntu with conhost as the terminal.

I think you might be onto something with columns, but it's not as easy as specifying it. The output is still incorrect, but I get different incorrect output depending on what value I pass in.

Some more screenshots:

With cols = 120:
![120-cols](https://user-images.githubusercontent.com/606923/29245665-44dbbd62-7fa7-11e7-9fc0-004c5b6ab9a0.png)

With cols = 80:
![80-cols](https://user-images.githubusercontent.com/606923/29245666-49d0eda6-7fa7-11e7-863a-4f64d6cc9c5d.png)

Line wrapping appears to be broken. Text in the right column is being cut off rather than wrapped to the next line, but only when it's not modified. Text in the left column wraps OK.

I thought that maybe it was an issue with the coloring, so I copied the text to see if perhaps it was the equivalent of black-on-black, but that was not the case. The text in the screenshots is all the text that is there.

**Comment by jeffkaufman** [comment]
This is very weird, thanks for the screenshots!  My guess at this point is it's some sort of bug in how conhost handles ascii color codes?  You could run `icdiff --cols=80  ... > tmp.txt` and then upload the output file, so I could verify that it looks reasonable on other systems?

My guess is that by playing around with this you could get to a minimal example that you could report to Microsoft and they could fix it, if that's really where the problem is.

**Comment by mqudsi** [comment]
Sure, here you go.

[out.txt](https://github.com/jeffkaufman/icdiff/files/1220239/out.txt)

**Comment by mqudsi** [comment]
In what is probably the silliest serious test, WSL/conhost has no problem rendering an animated nyancat:

![nyancat](https://user-images.githubusercontent.com/606923/29245962-14d048f0-7fb0-11e7-9e85-1f88867e2250.png)

**Comment by jeffkaufman** [comment]
Interesting: `cat out.txt` looks just as broken in iTerm2 and mac terminal:

<img width="656" alt="screen shot 2017-08-12 at 11 51 41 pm" src="https://user-images.githubusercontent.com/262566/29246004-38e8c3e4-7fb9-11e7-8878-af436b5ab181.png">

**Comment by jeffkaufman** [comment]
Ok, I think the problem is that your initial source files have `\r` in them, and I think this was actually already fixed on trunk with https://github.com/jeffkaufman/icdiff/commit/b5f8d11888b4e7ba65690b1f552f7a82a1c42b72

Can you try using the very latest icdiff and see if that fixes the problem? https://raw.githubusercontent.com/jeffkaufman/icdiff/master/icdiff

**Comment by mqudsi** [comment]
Ding ding ding!

![icdiff working](https://user-images.githubusercontent.com/606923/29251279-d148cee6-8017-11e7-8549-484a2e5f0bb4.png)

It's actually useable!

**Comment by jeffkaufman** [comment]
The literal `\r` is a difference between the two sides here, so probably good to show it? But if you don't like it you could use `--strip-trailing-cr`.

**Comment by mqudsi** [comment]
I guess the reason I was questioning it is that there's no difference between an actual, written out `\r` (two letters) and the `\r` 0x0A carriage return.

vim (and others) use ctrl+m, represented as `^M` for this reason.

**Comment by jeffkaufman** [comment]
Huh? I can't show things in a different font, this is the terminal. But `\r` on one side and a carriage return on the other will show up a change in the diff.

**Comment by mqudsi** [comment]
Of course not, I must have zeroed out for a minute.

Here's a screenshot of how vim shows it, note that in this case the new line was added:

![vim line endings](https://user-images.githubusercontent.com/606923/29252288-b1b838a2-8029-11e7-827e-7cae1a1cfe6e.png)

I imagine it was a purposeful decision to not highlight the ^M on their part, but I could see that going either way.

While `icdiff` would show the change between a literal `\r` and a `<CR>`, it wouldn't be apparent what the change was, would it? (I admit this is a corner case).

A tangential suggestion: if the file is entirely composed of `\r\n` line endings, you could hide the `\r` automatically (many command line editors will only show line endings in case of inconsistent line endings).

**Comment by mqudsi** [comment]
I just realized it doesn't make sense to show `\r` on changed lines if `\r` isn't shown anywhere else. It should either be shown on existing and new lines or on neither.

**Comment by jeffkaufman** [comment]
Fixed by 0e3a4cf
